tests/mocks/MockFileSystem.hh

```diff
@@ -0,0 +1,84 @@
+#ifndef RA_SERVICES_MOCK_FILESYSTEM_HH
+#define RA_SERVICES_MOCK_FILESYSTEM_HH
+#pragma once
+
+#include "services\IFileSystem.hh"
+#include "services\ServiceLocator.hh"
+#include "services\impl\StringTextReader.hh"
+#include "services\impl\StringTextWriter.hh"
+
+#include <map>
+
+namespace ra {
+namespace services {
+namespace mocks {
+
+class MockFileSystem : public IFileSystem
+{
+public:
+    MockFileSystem() : m_Override(this)
+    {
+    }
+
+    const std::wstring& BaseDirectory() const override { return m_sBaseDirectory; }
+    void SetBaseDirectory(const std::wstring& sBaseDirectory)
+    {
+        m_sBaseDirectory = sBaseDirectory;
+    }
+
+    bool DirectoryExists(const std::wstring& sDirectory) const override
+    {
+        assert(!"not implemented");
+        return false;
+    }
+
+    bool CreateDirectory(const std::wstring& sDirectory) const override
+    {
+        assert(!"not implemented");
+        return false;
+    }
+
+    void MockFile(const std::wstring& sPath, const std::string& sContents)
+    {
+        m_mFileContents[sPath] = sContents;
+    }
+
+    std::string& GetFileContents(const std::wstring& sPath)
+    {
+        auto pIter = m_mFileContents.find(sPath);
+        if (pIter != m_mFileContents.end())
+            return pIter->second;
+
+        static std::string sEmpty;
+        return sEmpty;
+    }
+
+    std::unique_ptr<TextReader> OpenTextFile(const std::wstring& sPath) const override
+    {
+        auto pIter = m_mFileContents.find(sPath);
+        if (pIter == m_mFileContents.end())
+            return std::unique_ptr<TextReader>();
+
+        auto pReader = std::make_unique<ra::services::impl::StringTextReader>(pIter->second);
+        return std::unique_ptr<TextReader>(pReader.release());
+    }
+
+    std::unique_ptr<TextWriter> CreateTextFile(const std::wstring& sPath) const override
+    {
+        std::string& sContents = m_mFileContents[sPath];
```

Comment:
Assignment with array indexing is a possible security issue. It may cause a CPU to be vulnerable to Spectre. It won't show up as a warning unless /Wall is specified but it's  too much of a hassle to turn it on without a precompiled header to disable all the warnings for library files.

Reply:
using the indexer allows for inserting or replacing without having to call find first. what's your proposed alternative?

Reply:
`at` doesn't work. from the documentation:

>  If no such element exists, an exception of type std::out_of_range is thrown.

the indexer will create the element if it doesn't exist.

Reply:
Alright dont worry about it. I'll do more research about it and address it later.

Reply:
If you dont care about order use std::unordered_map. Its find and at methods are constant on average but linear in the worst case. Std::map versions are logrithimic.

Reply:
thanks for the reminder. in this particular case, it's a mock for unit tests, so the collection shouldn't contain more than a handful of items. I don't expect a noticeable difference using either.